src/mlpack/methods/ann/augmented/tree_memory_impl.hpp

```diff
@@ -0,0 +1,158 @@
+{
+  assert(0 <= index && index < memorySize);
+  return memory.col(actualMemorySize - 1 + index);
+}
+
+template<typename T, typename J, typename W>
+inline arma::subview_col<T> TreeMemory<T, J, W>::Leaf(size_t index)
+{
+  assert(0 <= index && index < memorySize);
+  return memory.col(actualMemorySize - 1 + index);
+}
+
+template<typename T, typename J, typename W>
+inline arma::Col<T> TreeMemory<T, J, W>::Cell(size_t index) const
+{
+  assert(0 <= index && index < memory.size());
+  return memory.col(index);
+}
+
+template<typename T, typename J, typename W>
+inline arma::subview_col<T> TreeMemory<T, J, W>::Cell(size_t index)
+{
+  assert(0 <= index && index < memory.size());
+  return memory.col(index);
+}
+
+template<typename T, typename J, typename W>
+inline size_t TreeMemory<T, J, W>::Root() const
+{
+  return 0;
+}
+
+template<typename T, typename J, typename W>
+inline size_t TreeMemory<T, J, W>::Left(size_t origin) const
+{
+  return (origin << 1) + 1;
+}
+
+template<typename T, typename J, typename W>
+inline size_t TreeMemory<T, J, W>::Right(size_t origin) const
+{
+  return (origin << 1) + 2;
+}
+
+template<typename T, typename J, typename W>
+inline size_t TreeMemory<T, J, W>::LeafIndex(size_t leafPos) const
+{
+  assert(0 <= leafPos && leafPos < memorySize);
+  return actualMemorySize - 1 + leafPos;
+}
+
+template<typename T, typename J, typename W>
+inline size_t TreeMemory<T, J, W>::Parent(size_t child) const
+{
+  if (child == 0) return actualMemorySize;
+  return ((child + 1) >> 1) - 1;
+}
+
+template<typename T, typename J, typename W>
+void TreeMemory<T, J, W>::Initialize(arma::Mat<T>& leafValues)
+{
+  assert(leafValues.n_cols <= memorySize);
+  // First, write in the leaf nodes.
+  memory.cols(actualMemorySize - 1,
+              actualMemorySize - 1 + leafValues.n_cols - 1) = leafValues;
+  if (actualMemorySize < 2) return;
+  size_t lastWrittenIdx = actualMemorySize - 1 + leafValues.size() - 1;
+  // After that, write into inner nodes as prescribed by joinFunction.
+  for (size_t i = actualMemorySize - 2; i != Root(); --i)
+  {
+    size_t l = Left(i), r = Right(i);
+    assert(l <= r);
+    if (r > lastWrittenIdx) continue;
+    arma::Mat<T> newCellValue;
+    joinFunction.Predict(Stack(Cell(l), Cell(r)), newCellValue);
+    Cell(i) = newCellValue;
+  }
+}
+
+template<typename T, typename J, typename W>
+void TreeMemory<T, J, W>::Update(size_t pos, arma::Col<T> el)
+{
+  assert(pos >= 0 && pos < memorySize);
+  size_t start = actualMemorySize - 1 + pos;
+  arma::Mat<T> newCellValue;
+  writeFunction.Predict(Stack(Cell(start), el), newCellValue);
+  Cell(start) = newCellValue;
+  while (start != Root())
+  {
+    start = Parent(start);
+    size_t l = Left(start), r = Right(start);
+    joinFunction.Predict(Stack(Cell(l), Cell(r)), newCellValue);
+    Cell(start) = newCellValue;
+  }
+}
+
+template<typename T, typename J, typename W>
+arma::Mat<T> TreeMemory<T, J, W>::Stack(arma::Mat<T> left, arma::Mat<T> right) const
```

Comment:
Well, `Stack` is used for *arbitrary* columns, not necessarily the ones from the memory:

```C++
writeFunction.Predict(Stack(Cell(start), el), newCellValue);
```

Rewrote the function with const-reference parameters, since we don't really need to modify them or copy them inside a function.